lib/irq.c

```diff
@@ -243,21 +253,38 @@ irqs_set_data_eventfd(vfu_ctx_t *vfu_ctx, struct vfio_irq_set *irq_set,
     assert(data != NULL);
     for (i = irq_set->start, j = 0; i < (irq_set->start + irq_set->count);
          i++, j++) {
-        efd = vfu_ctx->irqs->efds[i];
+        if (irq_set->index == VFIO_PCI_ERR_IRQ_INDEX)
+            efd = vfu_ctx->irqs->err_efd;
+        else if (irq_set->index == VFIO_PCI_REQ_IRQ_INDEX)
+            efd = vfu_ctx->irqs->req_efd;
+        else
+            efd = vfu_ctx->irqs->efds[i];
         if (efd >= 0) {
             if (close(efd) == -1) {
                 vfu_log(vfu_ctx, LOG_DEBUG, "failed to close IRQ fd %d: %m", efd);
             }
-
-            vfu_ctx->irqs->efds[i] = -1;
+            if (irq_set->index == VFIO_PCI_ERR_IRQ_INDEX)
+                vfu_ctx->irqs->err_efd = -1;
+            else if (irq_set->index == VFIO_PCI_REQ_IRQ_INDEX)
+                vfu_ctx->irqs->req_efd = -1;
+            else
+                vfu_ctx->irqs->efds[i] = -1;
         }
         assert(data[j] >= 0);
         /*
          * We've already checked in handle_device_set_irqs that
          * nr_fds == irq_set->count.
          */
-        vfu_ctx->irqs->efds[i] = consume_fd(data, irq_set->count, j);
-        vfu_log(vfu_ctx, LOG_DEBUG, "event fd[%d]=%d", i, vfu_ctx->irqs->efds[i]);
+        if (irq_set->index == VFIO_PCI_ERR_IRQ_INDEX) {
```

Comment:
please just consume_fd() once into a local var, then set as appropriate

Reply:
If irq_set->index is equal to VFIO_PCI_ERR_IRQ_INDEX/VFIO_PCI_REQ_IRQ_INDEX, irq_set->count must be 1. So consume_fd() only is called once. I think this check can be added in device_set_irqs_validate?

Reply:
What I mean is, you are calling it on line :279, line :282 and line :285. I'm saying it would be simpler if you did this:

```
    efd = consume_fd(data, irq_set->count, j);
    switch (irq_set->index) {
    case VFIO_PCI_ERR_IRQ_INDEX:
           vfu_ctx->irqs->err_efd = efd;
           vfu_log(...
           break;
    ...
```

Reply:
Ok, sorry for the misunderstanding. I will modify it.